src/postgres.rs

```diff
@@ -0,0 +1,275 @@
+                    timestamp bigint not null,
+                    payload jsonb,
+                    PRIMARY KEY (source, psn)
+                )
+            "#,
+            self.table
+        );
+        let create_table_statement = client.prepare(&*create_table_sql).await?;
+        client.execute(&create_table_statement, &[]).await?;
+        info!("Created table");
+
+        let sql = format!(
+            r#"INSERT INTO {} (source, semantics, name, seq, psn, timestamp, payload)
+                 SELECT * FROM unnest($1,$2,$3,$4,$5,$6,$7)
+                 ON CONFLICT DO NOTHING"#,
+            &self.table
+        );
+        let types = vec![
+            Type::TEXT_ARRAY,
+            Type::TEXT_ARRAY,
+            Type::TEXT_ARRAY,
+            Type::INT8_ARRAY,
+            Type::INT8_ARRAY,
+            Type::INT8_ARRAY,
+            Type::JSONB_ARRAY,
+        ];
+        let insert_stmt = client.prepare_typed(&sql, &types).await?;
+
+        let conn = PostgresConnection {
+            client,
+            insert_stmt,
+            table: self.table.clone(),
+        };
+
+        Ok(conn)
+    }
+}
+
+#[async_trait]
+impl DbConnection for PostgresConnection {
+    async fn insert(&self, items: Vec<Event<Payload>>) -> Result<()> {
+        let start = Instant::now();
+
+        let num_rows = items.len();
+        debug!("Preparing {} events", num_rows);
+        let rows: DbEventVec = (&items).into();
+        let rows_suffix = if num_rows > 1 { "s" } else { "" };
+
+        let mut sources = rows.sources.clone();
+        sources.sort_unstable();
+        sources.dedup();
+        let sources_suffix = if sources.len() > 1 { "s" } else { "" };
+        let sources = sources.join(", ");
+
+        debug!(
+            "About to write {} record{} into DB. Source{}: {}",
+            num_rows, rows_suffix, sources_suffix, sources,
+        );
+        self.client
+            .execute(
+                &self.insert_stmt,
+                // Make sure that the order of the fields here matches the INSERT statement in the connect() method above
+                &[
+                    &rows.sources,
+                    &rows.semantics,
+                    &rows.names,
+                    &rows.lamports,
+                    &rows.offsets,
+                    &rows.timestamps,
+                    &rows.payloads,
+                ],
+            )
+            .await?;
+
+        let elapsed = start.elapsed().as_millis();
+        info!(
+            "Wrote {} record{} in {} ms ({} records/sec). Source{}: {}",
+            num_rows,
+            rows_suffix,
+            elapsed,
+            num_rows as u128 * 1_000 / elapsed,
+            sources_suffix,
+            sources,
+        );
+
+        Ok(())
+    }
+
+    async fn get_offsets(&self) -> Result<OffsetMap> {
+        info!("Querying initial offset map");
+        let sql = format!(
+            "SELECT source, MAX(psn) FROM {} GROUP BY source",
```

Comment:
This will work nicely for the BAG use-case, one downside is that insertion progress is only reported whenever there is something to insert — skipping over lots of uninteresting events will not be reflected (as opposed to the approach with a separate table).

Reply:
That is true (I noticed it and removed part of the misleading message). I'd rather keep it simple for now and perhaps make another PR with that change and also perhaps a change to the table schema to reflect our new terminology (and also figure out how to handle tags in EventService v2)